File: mytravelstatus/mobile/android/app/src/main/java/com/mytravelstatus/app/service/AnalyticsManager.kt
/**
 * AnalyticsManager.kt
 *
 * Privacy-focused analytics manager.
 * - Analytics are OFF by default (opt-in only)
 * - No sensitive data is ever tracked (no locations, no PII)
 * - All properties are sanitized before recording
 * - User can opt-out at any time via Settings
 *
 * Events tracked (when enabled):
 * - App lifecycle (launch, background)
 * - Screen views (screen names only, no content)
 * - Feature usage (counts only, no details)
 * - Permission responses (type only)
 * - Errors (type only, sanitized)
 *
 * Events NOT tracked:
 * - Location data (coordinates or country names)
 * - Trip details (dates, countries, notes)
 * - User identifiers (email, name)
 * - Device identifiers
 * - Photo metadata
 * - Calendar content
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

package com.mytravelstatus.app.service

import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date

/**
 * Analytics event types
 */
enum class AnalyticsEvent(val eventName: String) {
    // App Lifecycle
    APP_LAUNCHED("app_launched"),
    APP_BACKGROUNDED("app_backgrounded"),
    APP_FOREGROUNDED("app_foregrounded"),

    // Authentication
    LOGIN_STARTED("login_started"),
    LOGIN_COMPLETED("login_completed"),
    LOGIN_FAILED("login_failed"),
    LOGOUT("logout"),

    // Navigation
    SCREEN_VIEWED("screen_viewed"),
    TAB_SELECTED("tab_selected"),

    // Feature Usage
    TRIP_ADDED("trip_added"),
    TRIP_DELETED("trip_deleted"),
    MANUAL_CHECK_IN("manual_check_in"),
    PHOTO_IMPORT_STARTED("photo_import_started"),
    PHOTO_IMPORT_COMPLETED("photo_import_completed"),
    CALENDAR_IMPORT_STARTED("calendar_import_started"),
    CALENDAR_IMPORT_COMPLETED("calendar_import_completed"),
    DATA_EXPORT_REQUESTED("data_export_requested"),
    ACCOUNT_DELETION_REQUESTED("account_deletion_requested"),

    // Permissions
    PERMISSION_REQUESTED("permission_requested"),
    PERMISSION_GRANTED("permission_granted"),
    PERMISSION_DENIED("permission_denied"),

    // Subscriptions
    SUBSCRIPTION_VIEW_OPENED("subscription_view_opened"),
    SUBSCRIPTION_PURCHASE_STARTED("subscription_purchase_started"),
    SUBSCRIPTION_PURCHASE_COMPLETED("subscription_purchase_completed"),
    SUBSCRIPTION_PURCHASE_FAILED("subscription_purchase_failed"),
    SUBSCRIPTION_RESTORED("subscription_restored"),

    // Errors
    ERROR_OCCURRED("error_occurred"),
    SYNC_FAILED("sync_failed"),

    // Settings
    ANALYTICS_OPT_IN("analytics_opt_in"),
    ANALYTICS_OPT_OUT("analytics_opt_out"),
    BACKGROUND_LOCATION_ENABLED("background_location_enabled"),
    BACKGROUND_LOCATION_DISABLED("background_location_disabled")
}

/**
 * Analytics screen identifiers
 */
enum class AnalyticsScreen(val screenName: String) {
    HOME("home"),
    TRIPS("trips"),
    PASSPORT_CONTROL("passport_control"),
    PHOTO_IMPORT("photo_import"),
    CALENDAR_IMPORT("calendar_import"),
    SETTINGS("settings"),
    PRIVACY_SETTINGS("privacy_settings"),
    SUBSCRIPTION("subscription"),
    LOCATION_EDUCATION("location_education"),
    LOGIN("login"),
    ACCOUNT("account")
}

object AnalyticsManager {
    private const val TAG = "AnalyticsManager"

    private val _isEnabled = MutableStateFlow(false)
    val isEnabled: StateFlow<Boolean> = _isEnabled.asStateFlow()

    // Event buffer for when analytics is enabled but not yet initialized
    private val eventBuffer = mutableListOf<AnalyticsLogEntry>()

    // Maximum events to buffer
    private const val MAX_BUFFER_SIZE = 100

    /**
     * Initialize analytics (but don't enable yet)
     */
    fun initialize() {
        // Load enabled state from privacy settings
        _isEnabled.value = PrivacySettings.analyticsEnabled.value

        Log.d(TAG, "Analytics initialized, enabled: ${_isEnabled.value}")
    }

    /**
     * Enable analytics (user opted in)
     */
    fun enable() {
        _isEnabled.value = true
        PrivacySettings.setAnalyticsEnabled(true)
        track(AnalyticsEvent.ANALYTICS_OPT_IN)
        Log.d(TAG, "Analytics enabled")
    }

    /**
     * Disable analytics (user opted out)
     */
    fun disable() {
        track(AnalyticsEvent.ANALYTICS_OPT_OUT)
        _isEnabled.value = false
        PrivacySettings.setAnalyticsEnabled(false)
        eventBuffer.clear()
        Log.d(TAG, "Analytics disabled")
    }

    /**
     * Track an event
     * @param event The event to track
     * @param properties Optional properties (will be sanitized)
     */
    fun track(event: AnalyticsEvent, properties: Map<String, String>? = null) {
        if (!_isEnabled.value) {
            Log.d(TAG, "Analytics disabled, skipping event: ${event.eventName}")
            return
        }

        val sanitizedProperties = properties?.let { sanitizeProperties(it) }

        val entry = AnalyticsLogEntry(
            eventName = event.eventName,
            properties = sanitizedProperties,
            timestamp = Date()
        )

        // Buffer the event
        if (eventBuffer.size >= MAX_BUFFER_SIZE) {
            eventBuffer.removeAt(0) // Remove oldest
        }
        eventBuffer.add(entry)

        Log.d(TAG, "Tracked event: ${event.eventName}, properties: $sanitizedProperties")

        // TODO: Send to analytics service if configured
        // For now, events are just logged locally
    }

    /**
     * Track a screen view
     * @param screen The screen being viewed
     */
    fun trackScreen(screen: AnalyticsScreen) {
        track(AnalyticsEvent.SCREEN_VIEWED, mapOf("screen" to screen.screenName))
    }

    /**
     * Track a permission request/response
     * @param permissionType The type of permission (location, camera, etc.)
     * @param granted Whether permission was granted
     */
    fun trackPermission(permissionType: String, granted: Boolean) {
        val event = if (granted) AnalyticsEvent.PERMISSION_GRANTED else AnalyticsEvent.PERMISSION_DENIED
        track(event, mapOf("permission_type" to permissionType))
    }

    /**
     * Track an error (sanitized, no PII)
     * @param errorType The type of error
     */
    fun trackError(errorType: String) {
        // Sanitize error type - only allow alphanumeric and underscores
        val sanitizedType = errorType.replace(Regex("[^a-zA-Z0-9_]"), "_").take(50)
        track(AnalyticsEvent.ERROR_OCCURRED, mapOf("error_type" to sanitizedType))
    }

    /**
     * Export analytics data for user data export
     */
    fun exportData(): Map<String, Any> {
        return mapOf(
            "analyticsEnabled" to _isEnabled.value,
            "eventsTracked" to if (_isEnabled.value) {
                eventBuffer.map { entry ->
                    mapOf(
                        "event" to entry.eventName,
                        "properties" to (entry.properties ?: emptyMap()),
                        "timestamp" to entry.timestamp.time
                    )
                }
            } else {
                "Analytics disabled - no data collected"
            }
        )
    }

    /**
     * Clear all analytics data (for account deletion)
     */
    fun clearAllData() {
        eventBuffer.clear()
        _isEnabled.value = false
    }

    // MARK: - Private Methods

    /**
     * Sanitize properties to ensure no sensitive data is tracked
     * - Remove any potential PII patterns
     * - Limit string lengths
     * - Only allow whitelisted property keys
     */
    private fun sanitizeProperties(properties: Map<String, String>): Map<String, String> {
        val allowedKeys = setOf(
            "screen", "tab", "permission_type", "error_type",
            "import_count", "subscription_id", "source"
        )

        return properties
            .filterKeys { it in allowedKeys }
            .mapValues { (_, value) -> sanitizeValue(value) }
    }

    /**
     * Sanitize a single value
     */
    private fun sanitizeValue(value: String): String {
        // Remove potential PII patterns
        var sanitized = value

        // Remove email patterns
        sanitized = sanitized.replace(Regex("[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}"), "[EMAIL]")

        // Remove potential phone numbers
        sanitized = sanitized.replace(Regex("\\+?[0-9]{10,}"), "[PHONE]")

        // Remove potential coordinates
        sanitized = sanitized.replace(Regex("-?\\d+\\.\\d{4,}"), "[COORD]")

        // Limit length
        return sanitized.take(100)
    }
}

/**
 * Internal analytics log entry
 */
private data class AnalyticsLogEntry(
    val eventName: String,
    val properties: Map<String, String>?,
    val timestamp: Date
)

File: mytravelstatus/mobile/ios/MyTravelStatus/Services/AnalyticsManager.swift
/**
 * AnalyticsManager.swift
 *
 * Privacy-focused analytics implementation.
 * - Collects only anonymous, non-identifying events
 * - No precise location data
 * - No photo metadata
 * - No visa/tax/financial fields
 * - User opt-in required (disabled by default)
 *
 * @package MyTravelStatus
 * @since   1.0.0
 */

import Foundation

/// Analytics event types - carefully curated to avoid sensitive data
enum AnalyticsEvent: String {
    // App lifecycle
    case appLaunched = "app_launched"
    case appBackgrounded = "app_backgrounded"
    case appForegrounded = "app_foregrounded"

    // Navigation (no specific content)
    case screenViewed = "screen_viewed"
    case tabSelected = "tab_selected"

    // Feature usage (counts only, no content)
    case tripAdded = "trip_added"
    case tripEdited = "trip_edited"
    case tripDeleted = "trip_deleted"
    case manualCheckIn = "manual_check_in"

    // Permission flows (granted/denied only)
    case locationPermissionRequested = "location_permission_requested"
    case locationPermissionGranted = "location_permission_granted"
    case locationPermissionDenied = "location_permission_denied"
    case notificationPermissionRequested = "notification_permission_requested"
    case notificationPermissionGranted = "notification_permission_granted"
    case notificationPermissionDenied = "notification_permission_denied"

    // Import features (count only, no content)
    case photoImportStarted = "photo_import_started"
    case photoImportCompleted = "photo_import_completed"
    case calendarImportStarted = "calendar_import_started"
    case calendarImportCompleted = "calendar_import_completed"

    // Sync events (status only)
    case syncStarted = "sync_started"
    case syncCompleted = "sync_completed"
    case syncFailed = "sync_failed"

    // Subscription events
    case subscriptionViewed = "subscription_viewed"
    case subscriptionStarted = "subscription_started"
    case subscriptionRestored = "subscription_restored"
    case subscriptionCancelled = "subscription_cancelled"

    // Errors (type only, no details)
    case errorOccurred = "error_occurred"

    // Privacy actions
    case analyticsOptIn = "analytics_opt_in"
    case analyticsOptOut = "analytics_opt_out"
    case dataExportRequested = "data_export_requested"
    case accountDeletionRequested = "account_deletion_requested"
}

/// Screen names for analytics (no sensitive screens)
enum AnalyticsScreen: String {
    case home = "home"
    case trips = "trips"
    case passportControl = "passport_control"
    case settings = "settings"
    case subscription = "subscription"
    case privacy = "privacy"
    case photoImport = "photo_import"
    case calendarImport = "calendar_import"
    case locationEducation = "location_education"
    case onboarding = "onboarding"
}

@MainActor
class AnalyticsManager: ObservableObject {
    static let shared = AnalyticsManager()

    // MARK: - Properties

    @Published private(set) var isEnabled: Bool = false
    private var sessionId: String = UUID().uuidString
    private var eventQueue: [QueuedEvent] = []
    private let maxQueueSize = 100

    // Event batch upload (if using server-side analytics)
    private var uploadTimer: Timer?

    // MARK: - Initialization

    private init() {
        // Analytics is OFF by default - respects user privacy
        isEnabled = false
    }

    // MARK: - Public Methods

    /// Update analytics enabled state (called by PrivacySettings)
    func updateEnabled(_ enabled: Bool) {
        isEnabled = enabled

        if enabled {
            startSession()
        } else {
            endSession()
            clearQueue()
        }
    }

    /// Track an event (only if analytics is enabled)
    func track(_ event: AnalyticsEvent, properties: [String: String]? = nil) {
        guard isEnabled else { return }

        // Sanitize properties to ensure no sensitive data
        let sanitizedProperties = sanitizeProperties(properties)

        let queuedEvent = QueuedEvent(
            event: event.rawValue,
            properties: sanitizedProperties,
            timestamp: Date(),
            sessionId: sessionId
        )

        eventQueue.append(queuedEvent)

        // Trim queue if too large
        if eventQueue.count > maxQueueSize {
            eventQueue.removeFirst(eventQueue.count - maxQueueSize)
        }

        #if DEBUG
        print("[Analytics] \(event.rawValue): \(sanitizedProperties ?? [:])")
        #endif
    }

    /// Track screen view
    func trackScreen(_ screen: AnalyticsScreen) {
        track(.screenViewed, properties: ["screen": screen.rawValue])
    }

    /// Track error (type only, no details)
    func trackError(_ errorType: String) {
        // Only track error type, not message (could contain sensitive data)
        let sanitizedType = errorType
            .replacingOccurrences(of: "[0-9]", with: "", options: .regularExpression)
            .prefix(50)

        track(.errorOccurred, properties: ["type": String(sanitizedType)])
    }

    // MARK: - Private Methods

    private func startSession() {
        sessionId = UUID().uuidString

        // Start periodic upload if using server-side analytics
        // uploadTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
        //     self?.uploadEvents()
        // }
    }

    private func endSession() {
        uploadTimer?.invalidate()
        uploadTimer = nil
    }

    private func clearQueue() {
        eventQueue.removeAll()
    }

    /// Sanitize properties to remove any potential sensitive data
    private func sanitizeProperties(_ properties: [String: String]?) -> [String: String]? {
        guard let properties = properties else { return nil }

        var sanitized: [String: String] = [:]

        // Allowed property keys (whitelist approach)
        let allowedKeys: Set<String> = [
            "screen", "tab", "source", "count", "status", "type",
            "platform", "version", "error_type"
        ]

        for (key, value) in properties {
            // Only include whitelisted keys
            guard allowedKeys.contains(key.lowercased()) else { continue }

            // Sanitize value
            var sanitizedValue = value

            // Remove anything that looks like an email
            sanitizedValue = sanitizedValue.replacingOccurrences(
                of: "[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}",
                with: "[REDACTED]",
                options: .regularExpression
            )

            // Remove anything that looks like coordinates
            sanitizedValue = sanitizedValue.replacingOccurrences(
                of: "-?\\d+\\.\\d{4,}",
                with: "[REDACTED]",
                options: .regularExpression
            )

            // Truncate long values
            sanitizedValue = String(sanitizedValue.prefix(100))

            sanitized[key] = sanitizedValue
        }

        return sanitized.isEmpty ? nil : sanitized
    }

    // MARK: - Event Upload (Optional Server-Side)

    private func uploadEvents() {
        guard !eventQueue.isEmpty else { return }

        // If you want to send to a server:
        // let events = eventQueue
        // eventQueue.removeAll()
        // Task {
        //     try await APIClient.shared.uploadAnalytics(events)
        // }

        // For now, events are just logged locally
        #if DEBUG
        print("[Analytics] Queue has \(eventQueue.count) events")
        #endif
    }

    // MARK: - Data Export

    /// Export analytics data for user data export request
    func exportData() -> [String: Any] {
        return [
            "enabled": isEnabled,
            "sessionId": sessionId,
            "eventCount": eventQueue.count,
            "events": eventQueue.map { event in
                [
                    "event": event.event,
                    "timestamp": ISO8601DateFormatter().string(from: event.timestamp),
                    "properties": event.properties ?? [:]
                ]
            }
        ]
    }
}

// MARK: - Models

private struct QueuedEvent {
    let event: String
    let properties: [String: String]?
    let timestamp: Date
    let sessionId: String
}
